{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'run': 'LH00336_0006', 'level': 'debug'}"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import logging\n",
    "import requests\n",
    "import json\n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "sys.path.append(os.path.dirname(os.path.dirname(os.getcwd())))\n",
    "from lib.nanuq import Nanuq\n",
    "from lib.gapp import Phenotips\n",
    "from lib.gapp import BSSH\n",
    "from lib.samplesheet import SampleSheet\n",
    "\n",
    "nq    = Nanuq()\n",
    "pho   = Phenotips()\n",
    "bssh  = BSSH()\n",
    "\n",
    "class Args:\n",
    "    # Simulate an argparse instance\n",
    "    #\n",
    "    def __init__(self, run='LH00336_0006', level='info'):\n",
    "        self.run   = run\n",
    "        self.level = level\n",
    "\n",
    "args = Args(level='debug')\n",
    "vars(args)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Sample_ID',\n",
       " 'ReferenceGenomeDir',\n",
       " 'VariantCallingMode',\n",
       " 'QcCoverage1BedFile',\n",
       " 'QcCoverage2BedFile',\n",
       " 'QcCoverage3BedFile',\n",
       " 'QcCrossContaminationVcfFile']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sheet = SampleSheet('D:\\\\HSJ\\\\Projects\\\\PRAGMatIQ\\\\Runs\\\\Seq_10B_PRAG_20240130.csv')\n",
    "sheet.sections['DragenGermline_Data'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['[Header]',\n",
       " 'FileFormatVersion,2',\n",
       " 'RunName,LH00336_0009',\n",
       " 'InstrumentPlatform,NovaSeq',\n",
       " 'IndexOrientation,Forward',\n",
       " '',\n",
       " '[Reads]',\n",
       " 'Read1Cycles,151',\n",
       " 'Read2Cycles,151',\n",
       " 'Index1Cycles,10',\n",
       " 'Index2Cycles,10',\n",
       " '',\n",
       " '[BCLConvert_Settings]',\n",
       " 'SoftwareVersion,3.8.4',\n",
       " 'FastqCompressionFormat,gzip',\n",
       " '',\n",
       " '[BCLConvert_Data]',\n",
       " 'Lane,Sample_ID,Index,Index2,Sample_Project',\n",
       " '1,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '1,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '1,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '1,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '1,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '1,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '1,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '1,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '1,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '1,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '1,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '1,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '1,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '1,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '1,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '1,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '1,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '1,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '2,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '2,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '2,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '2,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '2,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '2,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '2,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '2,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '2,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '2,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '2,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '2,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '2,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '2,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '2,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '2,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '2,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '2,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '3,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '3,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '3,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '3,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '3,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '3,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '3,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '3,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '3,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '3,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '3,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '3,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '3,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '3,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '3,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '3,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '3,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '3,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '4,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '4,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '4,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '4,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '4,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '4,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '4,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '4,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '4,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '4,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '4,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '4,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '4,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '4,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '4,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '4,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '4,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '4,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '5,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '5,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '5,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '5,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '5,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '5,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '5,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '5,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '5,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '5,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '5,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '5,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '5,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '5,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '5,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '5,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '5,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '5,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '6,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '6,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '6,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '6,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '6,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '6,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '6,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '6,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '6,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '6,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '6,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '6,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '6,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '6,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '6,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '6,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '6,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '6,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '7,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '7,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '7,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '7,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '7,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '7,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '7,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '7,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '7,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '7,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '7,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '7,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '7,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '7,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '7,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '7,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '7,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '7,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '8,24159,TCTCATGATA,AACACGTGGA,PRAGMatIQ_CHUSJ',\n",
       " '8,24160,CGAGGCCAAG,GTGTTACCGG,PRAGMatIQ_CHUSJ',\n",
       " '8,24161,TTCACGAGAC,AGATTGTTAC,PRAGMatIQ_CHUSJ',\n",
       " '8,24162,GCGTGGATGG,TTGACCAATG,PRAGMatIQ_CHUSJ',\n",
       " '8,24163,TCTGGTATCC,CGTTGCTTAC,PRAGMatIQ_CHUSJ',\n",
       " '8,24164,CATTAGTGCG,TGACTACATA,PRAGMatIQ_CHUSJ',\n",
       " '8,24212,ACGGTCAGGA,CGGCCTCGTT,PRAGMatIQ_CUSM',\n",
       " '8,24213,GGCAAGCCAG,CAAGCATCCG,PRAGMatIQ_CUSM',\n",
       " '8,24214,TGTCGCTGGT,TCGTCTGACT,PRAGMatIQ_CUSM',\n",
       " '8,24215,ACCGTTACAA,CTCATAGCGA,PRAGMatIQ_CHUQ',\n",
       " '8,24216,TATGCCTTAC,AGACACATTA,PRAGMatIQ_CHUQ',\n",
       " '8,24217,ACTGGATCTA,TCGCCGCTAG,PRAGMatIQ_CHUQ',\n",
       " '8,24251,TGGTACCTAA,CATGAGTACT,PRAGMatIQ_CHUS',\n",
       " '8,24252,TTGGAATTCC,ACGTCAATAC,PRAGMatIQ_CHUS',\n",
       " '8,24253,CCTCTACATG,GATACCTCCT,PRAGMatIQ_CHUS',\n",
       " '8,24254,GGAGCGTGTA,ATCCGTAAGT,PRAGMatIQ_CHUSJ',\n",
       " '8,24255,TTACTCCACA,GACACCGATG,PRAGMatIQ_CHUSJ',\n",
       " '8,24256,AGTAGAAGTG,CTAGCGTCGA,PRAGMatIQ_CHUSJ',\n",
       " '',\n",
       " '[Cloud_Settings]',\n",
       " 'GeneratedVersion,1.1.0.202107141549',\n",
       " 'BsshApp,illumina-inc.basespace-dragen-analysis.1.2.1',\n",
       " '',\n",
       " '[Cloud_Data]',\n",
       " 'Sample_ID,ProjectName,LibraryName,LibraryPrepKitUrn,LibraryPrepKitName,IndexAdapterKitUrn,IndexAdapterKitName',\n",
       " '24159,PRAGMatIQ_CHUSJ,24159_TCTCATGATA_AACACGTGGA,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24160,PRAGMatIQ_CHUSJ,24160_CGAGGCCAAG_GTGTTACCGG,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24161,PRAGMatIQ_CHUSJ,24161_TTCACGAGAC_AGATTGTTAC,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24162,PRAGMatIQ_CHUSJ,24162_GCGTGGATGG_TTGACCAATG,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24163,PRAGMatIQ_CHUSJ,24163_TCTGGTATCC_CGTTGCTTAC,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24164,PRAGMatIQ_CHUSJ,24164_CATTAGTGCG_TGACTACATA,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24212,PRAGMatIQ_CUSM,24212_ACGGTCAGGA_CGGCCTCGTT,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24213,PRAGMatIQ_CUSM,24213_GGCAAGCCAG_CAAGCATCCG,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24214,PRAGMatIQ_CUSM,24214_TGTCGCTGGT_TCGTCTGACT,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24215,PRAGMatIQ_CHUQ,24215_ACCGTTACAA_CTCATAGCGA,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24216,PRAGMatIQ_CHUQ,24216_TATGCCTTAC_AGACACATTA,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24217,PRAGMatIQ_CHUQ,24217_ACTGGATCTA_TCGCCGCTAG,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24251,PRAGMatIQ_CHUS,24251_TGGTACCTAA_CATGAGTACT,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24252,PRAGMatIQ_CHUS,24252_TTGGAATTCC_ACGTCAATAC,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24253,PRAGMatIQ_CHUS,24253_CCTCTACATG_GATACCTCCT,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24254,PRAGMatIQ_CHUSJ,24254_GGAGCGTGTA_ATCCGTAAGT,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24255,PRAGMatIQ_CHUSJ,24255_TTACTCCACA_GACACCGATG,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '24256,PRAGMatIQ_CHUSJ,24256_AGTAGAAGTG_CTAGCGTCGA,,Illumina_DNA_PCR-Free_Tagmentation_2,,,',\n",
       " '',\n",
       " '[CQGC_Data]',\n",
       " 'Sample_ID,Control']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "content = nq.get_samplesheet('LH00336_0009').text.splitlines()\n",
    "content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = []\n",
    "for line in content:\n",
    "    if line.startswith('['):\n",
    "        section = line\n",
    "    else:\n",
    "        if section.startswith('[DragenGermline_Data]'):\n",
    "            cols = line.split(',')\n",
    "            if not line.startswith('Sample_ID') and len(cols) > 1:\n",
    "                samples.append(cols[0])\n",
    "samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get authorization token from Emedgene\n",
    "# Please note - the Authorization header is only valid for a limited time, and\n",
    "# expires after 8H. In that case, any request made with an expired token will \n",
    "# return a 403 code. To resolve, re-do the Login procedure to get a new token.\n",
    "#\n",
    "url      = \"https://chusaintejustine.emedgene.com/api/auth/api_login/\"\n",
    "payload  = '{\"username\": \"cqgc.bioinfo.hsj@ssss.gouv.qc.ca\", \"password\": \"3175CoteSainte-Catherine\"}'\n",
    "headers  = {'Content-Type': 'application/json'}\n",
    "response = requests.request(\"POST\", url, headers=headers, data=payload)\n",
    "auth_header = response.json()[\"Authorization\"]\n",
    "auth_header\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get case for EMG634551172 (GM240123)\n",
    "#\n",
    "case_id = \"EMG634551172\"\n",
    "get_test_response = requests.get(f'https://chusaintejustine.emedgene.com/api/test/{case_id}/', json={}, headers={'Authorization': auth_header})\n",
    "# with open('D:\\\\HSJ\\\\Projects\\\\PRAGMatIQ\\\\emg_scripts\\\\emg_case_example-GM240123.json', 'w') as fh:\n",
    "#     json.dump(get_test_response.json(), fh, indent=4)\n",
    "get_test_response.json().keys()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# Transfer files to aws\n",
    "# --profile \"emedgene\" or \"emedgene-eval\"\n",
    "aws s3 ls --profile emedgene-eval s3://cac1-prodca-emg-downloads/Ste_Justine_eval/upload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'24159': {'aligned_bases': 116701731183,\n",
       "  'aligned_bases_in_genome': 116701731183,\n",
       "  'aligned_bases_in_genome_pct': 100.0,\n",
       "  'aligned_reads': 813636857,\n",
       "  'aligned_reads_in_genome': 813636857,\n",
       "  'aligned_reads_in_genome_pct': 100.0,\n",
       "  'autosomal_median_coverage': 52.79,\n",
       "  'average_alignment_coverage_over_genome': 38.61,\n",
       "  'average_autosomal_coverage_over_genome': 40.86,\n",
       "  'average_chr_x_coverage_over_genome': 20.09,\n",
       "  'average_chr_y_coverage_over_genome': 17.58,\n",
       "  'average_mitochondrial_coverage_over_genome': 2981.83,\n",
       "  'bases_in_reference_genome': 3706390998,\n",
       "  'chr10_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr11_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr12_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr13_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr14_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr15_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr16_autosomal_median_coverage_ratio': 1.01,\n",
       "  'chr17_autosomal_median_coverage_ratio': 1.01,\n",
       "  'chr18_autosomal_median_coverage_ratio': 0.99,\n",
       "  'chr19_autosomal_median_coverage_ratio': 1.02,\n",
       "  'chr1_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr20_autosomal_median_coverage_ratio': 1.01,\n",
       "  'chr21_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr22_autosomal_median_coverage_ratio': 1.02,\n",
       "  'chr2_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr3_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr4_autosomal_median_coverage_ratio': 0.99,\n",
       "  'chr5_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr6_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr7_autosomal_median_coverage_ratio': 1.0,\n",
       "  'chr8_autosomal_median_coverage_ratio': 0.99,\n",
       "  'chr9_autosomal_median_coverage_ratio': 1.0,\n",
       "  'cnv_average_alignment_coverage_over_genome': 40.11,\n",
       "  'cnv_bases_in_reference_genome': 3706390998,\n",
       "  'cnv_coverage_uniformity': 0.14,\n",
       "  'cnv_number_of_alignment_records': 1107637280,\n",
       "  'cnv_number_of_amplifications': 200,\n",
       "  'cnv_number_of_deletions': 500,\n",
       "  'cnv_number_of_filtered_records_duplicates': 0,\n",
       "  'cnv_number_of_filtered_records_duplicates_pct': 0.0,\n",
       "  'cnv_number_of_filtered_records_mapq': 61973012,\n",
       "  'cnv_number_of_filtered_records_mapq_pct': 5.6,\n",
       "  'cnv_number_of_filtered_records_total': 77745548,\n",
       "  'cnv_number_of_filtered_records_total_pct': 7.02,\n",
       "  'cnv_number_of_filtered_records_unmapped': 15772536,\n",
       "  'cnv_number_of_filtered_records_unmapped_pct': 1.42,\n",
       "  'cnv_number_of_passing_amplifications': 125,\n",
       "  'cnv_number_of_passing_amplifications_pct': 62.5,\n",
       "  'cnv_number_of_passing_deletions': 61,\n",
       "  'cnv_number_of_passing_deletions_pct': 12.2,\n",
       "  'cnv_number_of_segments': 2070,\n",
       "  'cnv_number_of_target_intervals': 2441294,\n",
       "  'dragen_mapping_rate_mil_reads_second': 0.66,\n",
       "  'estimated_read_length': 144.8,\n",
       "  'estimated_sample_contamination': 0.0,\n",
       "  'insert_length_mean': 374.42,\n",
       "  'insert_length_median': 374.5,\n",
       "  'insert_length_standard_deviation': 164.46,\n",
       "  'mapped_bases_r1': 79417522264,\n",
       "  'mapped_bases_r2': 78877511626,\n",
       "  'mapped_reads': 1090686060,\n",
       "  'mapped_reads_pct': 98.57,\n",
       "  'mapped_reads_r1': 547002443,\n",
       "  'mapped_reads_r1_pct': 98.87,\n",
       "  'mapped_reads_r2': 543683617,\n",
       "  'mapped_reads_r2_pct': 98.27,\n",
       "  'mean_median_autosomal_coverage_ratio_over_genome': 0.97,\n",
       "  'median_autosomal_coverage_over_genome': 42.0,\n",
       "  'mismatched_bases_r1': 394539133,\n",
       "  'mismatched_bases_r1_excl_indels': 335557862,\n",
       "  'mismatched_bases_r1_excl_indels_pct': 0.42,\n",
       "  'mismatched_bases_r1_pct': 0.5,\n",
       "  'mismatched_bases_r2': 563574161,\n",
       "  'mismatched_bases_r2_excl_indels': 508371150,\n",
       "  'mismatched_bases_r2_excl_indels_pct': 0.64,\n",
       "  'mismatched_bases_r2_pct': 0.71,\n",
       "  'not_properly_paired_reads_discordant': 6677024,\n",
       "  'not_properly_paired_reads_discordant_pct': 0.6,\n",
       "  'number_of_duplicate_marked_reads': 228118324,\n",
       "  'number_of_duplicate_marked_reads_pct': 20.62,\n",
       "  'number_of_large_roh_gt_eq_3000000': 1,\n",
       "  'number_of_unique_reads_excl_duplicate_marked_reads': 878340272,\n",
       "  'number_of_unique_reads_excl_duplicate_marked_reads_pct': 79.38,\n",
       "  'number_unique_mapped_reads_excl_duplicates': 862567736,\n",
       "  'number_unique_mapped_reads_excl_duplicates_pct': 77.96,\n",
       "  'paired_reads_different_chromosomes_mapq_gt_eq_10': 1417634,\n",
       "  'paired_reads_different_chromosomes_mapq_gt_eq_10_pct': 0.13,\n",
       "  'paired_reads_itself_and_mate_mapped': 1078257912,\n",
       "  'paired_reads_itself_and_mate_mapped_pct': 97.45,\n",
       "  'paired_reads_mapped_to_different_chromosomes': 4226968,\n",
       "  'paired_reads_mapped_to_different_chromosomes_pct': 0.39,\n",
       "  'pct_of_genome_with_coverage_0x_1x': 4.51,\n",
       "  'pct_of_genome_with_coverage_0x_inf': 100.0,\n",
       "  'pct_of_genome_with_coverage_100x_inf': 0.19,\n",
       "  'pct_of_genome_with_coverage_10x_15x': 0.98,\n",
       "  'pct_of_genome_with_coverage_10x_inf': 94.24,\n",
       "  'pct_of_genome_with_coverage_15x_20x': 2.13,\n",
       "  'pct_of_genome_with_coverage_15x_inf': 93.27,\n",
       "  'pct_of_genome_with_coverage_1x_3x': 0.49,\n",
       "  'pct_of_genome_with_coverage_1x_inf': 95.49,\n",
       "  'pct_of_genome_with_coverage_20x_50x': 74.94,\n",
       "  'pct_of_genome_with_coverage_20x_inf': 91.14,\n",
       "  'pct_of_genome_with_coverage_3x_10x': 0.76,\n",
       "  'pct_of_genome_with_coverage_3x_inf': 95.0,\n",
       "  'pct_of_genome_with_coverage_50x100x': 16.01,\n",
       "  'pct_of_genome_with_coverage_50x_inf': 16.2,\n",
       "  'percent_snvs_in_large_roh_gt_eq_3000000': 0.031,\n",
       "  'ploidy_estimation': 'XY',\n",
       "  'properly_paired_reads': 1071580888,\n",
       "  'properly_paired_reads_pct': 96.85,\n",
       "  'provided_sex_chromosome_ploidy': 'NA',\n",
       "  'q30_bases': 148579145496,\n",
       "  'q30_bases_excl_dups_and_clipped_bases': 117497166217,\n",
       "  'q30_bases_pct': 92.74,\n",
       "  'q30_bases_r1': 75795475025,\n",
       "  'q30_bases_r1_pct': 94.83,\n",
       "  'q30_bases_r2': 72783670471,\n",
       "  'q30_bases_r2_pct': 90.65,\n",
       "  'qc_failed_reads': 0,\n",
       "  'qc_failed_reads_pct': 0.0,\n",
       "  'reads_with_indel_r1': 17649788,\n",
       "  'reads_with_indel_r1_pct': 3.23,\n",
       "  'reads_with_indel_r2': 16745965,\n",
       "  'reads_with_indel_r2_pct': 3.08,\n",
       "  'reads_with_mapq_010': 66332310,\n",
       "  'reads_with_mapq_010_pct': 6.0,\n",
       "  'reads_with_mapq_1020': 11679109,\n",
       "  'reads_with_mapq_1020_pct': 1.06,\n",
       "  'reads_with_mapq_2030': 9063028,\n",
       "  'reads_with_mapq_2030_pct': 0.82,\n",
       "  'reads_with_mapq_3040': 4673168,\n",
       "  'reads_with_mapq_3040_pct': 0.42,\n",
       "  'reads_with_mapq_40inf': 998938445,\n",
       "  'reads_with_mapq_40inf_pct': 90.28,\n",
       "  'reads_with_mapq_na_unmapped_reads': 15772536,\n",
       "  'reads_with_mapq_na_unmapped_reads_pct': 1.43,\n",
       "  'reads_with_mate_sequenced': 1106458596,\n",
       "  'reads_with_mate_sequenced_pct': 100.0,\n",
       "  'reads_without_mate_sequenced': 0,\n",
       "  'reads_without_mate_sequenced_pct': 0.0,\n",
       "  'secondary_alignments': 0,\n",
       "  'singleton_reads_itself_mapped_mate_unmapped': 12428148,\n",
       "  'singleton_reads_itself_mapped_mate_unmapped_pct': 1.12,\n",
       "  'soft_clipped_bases_r1': 560051195,\n",
       "  'soft_clipped_bases_r1_pct': 0.71,\n",
       "  'soft_clipped_bases_r2': 861567591,\n",
       "  'soft_clipped_bases_r2_pct': 1.09,\n",
       "  'supplementary_chimeric_alignments': 1178684,\n",
       "  'sv_number_of_breakend_pairs_pass': 784,\n",
       "  'sv_number_of_deletions_pass': 5349,\n",
       "  'sv_number_of_duplications_pass': 50,\n",
       "  'sv_number_of_insertions_pass': 6415,\n",
       "  'total_alignments': 1091864744,\n",
       "  'total_bases': 160213693193,\n",
       "  'total_bases_r1': 79926185600,\n",
       "  'total_bases_r2': 80287507593,\n",
       "  'total_input_reads': 1106458596,\n",
       "  'total_input_reads_pct': 100.0,\n",
       "  'uniformity_of_coverage_pct_gt_02mean_over_genome': 94.45,\n",
       "  'unmapped_reads': 15772536,\n",
       "  'unmapped_reads_pct': 1.43,\n",
       "  'variants_biallelic_pass': 5049520,\n",
       "  'variants_biallelic_pass_pct': 98.08,\n",
       "  'variants_chr_x_number_of_snps_over_genome_pass': 92159,\n",
       "  'variants_chr_x_snps_chr_y_snps_ratio_over_genome_pass': 94.91,\n",
       "  'variants_chr_y_number_of_snps_over_genome_pass': 971,\n",
       "  'variants_deletions_het_pass': 321012,\n",
       "  'variants_deletions_het_pass_pct': 6.24,\n",
       "  'variants_deletions_hom_pass': 155694,\n",
       "  'variants_deletions_hom_pass_pct': 3.02,\n",
       "  'variants_het_to_hom_ratio_pass': 1.76,\n",
       "  'variants_heterozygous_pass': 3214388,\n",
       "  'variants_homozygous_pass': 1823387,\n",
       "  'variants_in_dbsnp_pass': 0,\n",
       "  'variants_in_dbsnp_pass_pct': 0.0,\n",
       "  'variants_indels_het_pass': 27388,\n",
       "  'variants_indels_het_pass_pct': 0.53,\n",
       "  'variants_insertions_het_pass': 315752,\n",
       "  'variants_insertions_het_pass_pct': 6.13,\n",
       "  'variants_insertions_hom_pass': 157956,\n",
       "  'variants_insertions_hom_pass_pct': 3.07,\n",
       "  'variants_multiallelic_pass': 98851,\n",
       "  'variants_multiallelic_pass_pct': 1.92,\n",
       "  'variants_not_in_dbsnp_pass': 0,\n",
       "  'variants_not_in_dbsnp_pass_pct': 0.0,\n",
       "  'variants_number_of_samples': 1,\n",
       "  'variants_reads_processed': 850060564,\n",
       "  'variants_snp_transitions_pass': 3757466,\n",
       "  'variants_snp_transversions_pass': 1899623,\n",
       "  'variants_snps_pass': 4144821,\n",
       "  'variants_snps_pass_pct': 80.51,\n",
       "  'variants_ti_to_tv_ratio_pass': 1.98,\n",
       "  'variants_total_pass': 5148371,\n",
       "  'variants_total_pass_pct': 100.0,\n",
       "  'x_median_autosomal_median': 0.5,\n",
       "  'x_median_coverage': 26.23,\n",
       "  'y_median_autosomal_median': 0.51,\n",
       "  'y_median_coverage': 26.78}}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = '24159'\n",
    "samples_metrics = {}\n",
    "with open(f\"{sample}.metrics.json\", \"r\") as fh:\n",
    "    metrics = json.load(fh)['Attributes']['illumina_dragen_complete_v0_4']\n",
    "\n",
    "samples_metrics[sample] = metrics\n",
    "samples_metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aligned_bases</th>\n",
       "      <th>aligned_bases_in_genome</th>\n",
       "      <th>aligned_bases_in_genome_pct</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>aligned_reads_in_genome</th>\n",
       "      <th>aligned_reads_in_genome_pct</th>\n",
       "      <th>autosomal_median_coverage</th>\n",
       "      <th>average_alignment_coverage_over_genome</th>\n",
       "      <th>average_autosomal_coverage_over_genome</th>\n",
       "      <th>average_chr_x_coverage_over_genome</th>\n",
       "      <th>...</th>\n",
       "      <th>variants_snp_transversions_pass</th>\n",
       "      <th>variants_snps_pass</th>\n",
       "      <th>variants_snps_pass_pct</th>\n",
       "      <th>variants_ti_to_tv_ratio_pass</th>\n",
       "      <th>variants_total_pass</th>\n",
       "      <th>variants_total_pass_pct</th>\n",
       "      <th>x_median_autosomal_median</th>\n",
       "      <th>x_median_coverage</th>\n",
       "      <th>y_median_autosomal_median</th>\n",
       "      <th>y_median_coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>24159</th>\n",
       "      <td>116701731183</td>\n",
       "      <td>116701731183</td>\n",
       "      <td>100.0</td>\n",
       "      <td>813636857</td>\n",
       "      <td>813636857</td>\n",
       "      <td>100.0</td>\n",
       "      <td>52.79</td>\n",
       "      <td>38.61</td>\n",
       "      <td>40.86</td>\n",
       "      <td>20.09</td>\n",
       "      <td>...</td>\n",
       "      <td>1899623</td>\n",
       "      <td>4144821</td>\n",
       "      <td>80.51</td>\n",
       "      <td>1.98</td>\n",
       "      <td>5148371</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.5</td>\n",
       "      <td>26.23</td>\n",
       "      <td>0.51</td>\n",
       "      <td>26.78</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 201 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       aligned_bases  aligned_bases_in_genome  aligned_bases_in_genome_pct  \\\n",
       "24159   116701731183             116701731183                        100.0   \n",
       "\n",
       "       aligned_reads  aligned_reads_in_genome  aligned_reads_in_genome_pct  \\\n",
       "24159      813636857                813636857                        100.0   \n",
       "\n",
       "       autosomal_median_coverage  average_alignment_coverage_over_genome  \\\n",
       "24159                      52.79                                   38.61   \n",
       "\n",
       "       average_autosomal_coverage_over_genome  \\\n",
       "24159                                   40.86   \n",
       "\n",
       "       average_chr_x_coverage_over_genome  ...  \\\n",
       "24159                               20.09  ...   \n",
       "\n",
       "       variants_snp_transversions_pass  variants_snps_pass  \\\n",
       "24159                          1899623             4144821   \n",
       "\n",
       "       variants_snps_pass_pct  variants_ti_to_tv_ratio_pass  \\\n",
       "24159                   80.51                          1.98   \n",
       "\n",
       "       variants_total_pass  variants_total_pass_pct  \\\n",
       "24159              5148371                    100.0   \n",
       "\n",
       "       x_median_autosomal_median  x_median_coverage  \\\n",
       "24159                        0.5              26.23   \n",
       "\n",
       "       y_median_autosomal_median  y_median_coverage  \n",
       "24159                       0.51              26.78  \n",
       "\n",
       "[1 rows x 201 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_samples_metrics = pd.DataFrame.from_dict(samples_metrics, orient=\"index\")\n",
    "df_samples_metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_cols = ['aligned_reads', 'aligned_reads_in_genome_pct', 'cnv_coverage_uniformity']\n",
    "df = df_samples_metrics"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "dev",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
